core/slice.go

```diff
@@ -0,0 +1,647 @@
+		if err != nil {
+			return sl.nilHeader, err
+		}
+		sl.hc.chainHeadFeed.Send(ChainHeadEvent{Block: block})
+	} else {
+		sl.hc.chainSideFeed.Send(ChainSideEvent{Block: block})
+	}
+
+	// Upate the local pending header
+	pendingHeader, err := sl.miner.worker.GeneratePendingHeader(block)
+	if err != nil {
+		return sl.nilHeader, err
+	}
+
+	// Set the Location and time for the pending header
+	pendingHeader.SetLocation(common.NodeLocation)
+
+	return pendingHeader, nil
+}
+
+// PCRC previous coincidence reference check makes sure there are not any cyclic references in the graph and calculates new termini and the block terminus
+func (sl *Slice) pcrc(batch ethdb.Batch, header *types.Header, domTerminus common.Hash) (common.Hash, []common.Hash, error) {
+	nodeCtx := common.NodeLocation.Context()
+	location := header.Location()
+
+	isCoincident := sl.engine.HasCoincidentDifficulty(header)
+
+	log.Debug("PCRC:", "Parent Hash:", header.ParentHash(), "Number", header.Number, "Location:", header.Location())
+	termini := sl.hc.GetTerminiByHash(header.ParentHash())
+
+	if len(termini) != 4 {
+		return common.Hash{}, []common.Hash{}, errors.New("length of termini not equal to 4")
+	}
+
+	newTermini := make([]common.Hash, len(termini))
+	for i, terminus := range termini {
+		newTermini[i] = terminus
+	}
+
+	// Genesis escape for the domTerminus
+	if header.ParentHash(common.PRIME_CTX) == sl.config.GenesisHash {
+		domTerminus = sl.config.GenesisHash
+	}
+
+	// Set the subtermini
+	if nodeCtx != common.ZONE_CTX {
+		newTermini[location.SubLocation()] = header.Hash()
+	}
+
+	// Set the terminus
+	if nodeCtx == common.PRIME_CTX || isCoincident {
+		newTermini[3] = header.Hash()
+	} else {
+		newTermini[3] = termini[3]
+	}
+
+	// Check for a graph twist
+	if isCoincident {
+		if termini[3] != domTerminus {
+			return common.Hash{}, []common.Hash{}, errors.New("termini do not match, block rejected due to a twist")
+		}
+	}
+
+	//Save the termini
+	rawdb.WriteTermini(batch, header.Hash(), newTermini)
+
+	if nodeCtx == common.ZONE_CTX {
+		return common.Hash{}, newTermini, nil
+	}
+
+	return termini[location.SubLocation()], newTermini, nil
+}
+
+// HLCR Hierarchical Longest Chain Rule compares externTd to the currentHead Td and returns true if externTd is greater
+func (sl *Slice) hlcr(externTd *big.Int) bool {
+	currentTd := sl.hc.GetTdByHash(sl.hc.CurrentHeader().Hash())
+	log.Debug("HLCR:", "Header hash:", sl.hc.CurrentHeader().Hash(), "currentTd:", currentTd, "externTd:", externTd)
+	reorg := currentTd.Cmp(externTd) < 0
+	//TODO need to handle the equal td case
+	return reorg
+}
+
+// CalcTd calculates the TD of the given header using PCRC and CalcHLCRNetDifficulty.
+func (sl *Slice) calcTd(header *types.Header) (*big.Int, error) {
+	priorTd := sl.hc.GetTd(header.ParentHash(), header.NumberU64()-1)
+	if priorTd == nil {
+		return nil, consensus.ErrFutureBlock
+	}
+	Td := priorTd.Add(priorTd, header.Difficulty())
+	return Td, nil
```

Comment:
This is not hierarchical TD. Should it be?

Reply:
Yes it is and it is also only used in the case of non dom origin block. Also can you clarify what you mean by hierarchical TD.

Reply:
Just add a guard here to ensure this can't be called on a non-coincident block

Reply:
Added and returning error.